### PR TITLE
[Backport 2019.2] Fix bug that didn't populate colors when precision changed

### PR DESCRIPTION
**Pull request by samhardin**
### Purpose of this PR
There was a bug that didn't populate node colors downstream when an edge was added or removed if the Color Mode was in **Precision**.
**Broken Version:**
ColorModePrecisionBUG

**Fixed Version:**
ColorModePrecision

---
### Release Notes
N/A

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [x] Tested UI multi-edition + Undo/Redo
- [x] C# and shader warnings (supress shader cache to see them)

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

Launch Katana (Link on master here - update for your branch):
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=sg%2Fcolor-mode-minor-fix&automation-tools_branch=master&unity_branch=trunk&sort=1-asc

Alternative, launch Yamato (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Fcolor-mode-minor-fix

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High? Low

**Halo Effect**: None, Low, Medium, High? Low

---
### Comments to reviewers
N/A

### CODE REVIEW

**Comment by zadi0648** [comment]
Another:

1) Have precision be Half for the graph
2) Have a node ("node 1") with precision going into any node.
3) Create another node ("node 3") (keep it as inherit).
4) Connect "node 1" to "node 3"
> Bug: node 3 doesn't change color as it should
I checked Generated Code to confirm this isn't back end.
Note that it does not reproduce if you drag then create from the drag, as opposed to create then connect.

2ndConnect

**Comment by zadi0648** [comment]
Found a third issue while waiting for the guys:

For SubGraphs, things don't inherit if they are created first then connect to the SubGraph. So kind of like the above one, except it reproduces for the first edge.

In this picture, the right column shows the correct columns. The middle column however is bugged. They were created and then connected (and the SubGraphs had no other edges).
SubGraphs

**Comment by zadi0648** [comment]
And yes I confirmed that this reproduces in when using the 2019.2 😛 (I didn't notice this was a backport branch at first and was testing on trunk)

File: com.unity.shadergraph/Editor/Drawing/PreviewManager.cs [hunk 1]
```diff
@@ -27,6 +27,11 @@ class PreviewManager : IDisposable
         HashSet<AbstractMaterialNode> m_TimedNodes = new HashSet<AbstractMaterialNode>();
         bool m_RefreshTimedNodes;
 
+        public HashSet<AbstractMaterialNode> nodesToUpdate
```
[the remark this answers is not in the document]
Comment (samhardin) [reply com.unity.shadergraph/Editor/Drawing/PreviewManager.cs]: Definitely! I pulled that accessor out and am now just returning true if any nodes are ready to update.

File: com.unity.shadergraph/Editor/Drawing/Colors/IColorProvider.cs [hunk 2]
```diff
@@ -68,5 +74,13 @@ public virtual void ClearColor(IShaderNodeView nodeView)
                 nodeView.colorElement.RemoveFromClassList(ussClass);
             }
         }
+
+        public virtual void ClearAllColors(IShaderNodeView nodeView)
+        {
+            foreach(var type in ConcretePrecision.GetValues(typeof(ConcretePrecision)))
```
[the remark this answers is not in the document]
Comment (samhardin) [reply com.unity.shadergraph/Editor/Drawing/Colors/IColorProvider.cs]: So what I ended up doing is simply overriding the ClearColor function in PrecisionColors while taking out the ClearAllColors function all together.

File: com.unity.shadergraph/Editor/Drawing/Colors/ColorManager.cs [hunk 3]
```diff
@@ -54,7 +54,7 @@ public void SetNodesDirty(IEnumerable<IShaderNodeView> nodeViews)
                 }
             }
         }
-        
+
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Colors/ColorManager.cs]: Can you revert this whitespace-only change so the file isn't part of the PR please?

File: com.unity.shadergraph/Editor/Drawing/Colors/NoColors.cs [hunk 4]
```diff
@@ -18,5 +18,9 @@ public void ApplyColor(IShaderNodeView nodeView)
         public void ClearColor(IShaderNodeView nodeView)
         {
         }
+
+        public void ClearAllColors(IShaderNodeView nodeView)
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Colors/NoColors.cs]: This function is no longer needed, please remove so this file isn't part of the diff either.

File: com.unity.shadergraph/Editor/Drawing/PreviewManager.cs [hunk 5]
```diff
@@ -248,6 +248,11 @@ public void HandleGraphChanges()
                     m_RefreshTimedNodes = true;
                 }
             }
+
+            if (m_NodesToUpdate.Count > 0)
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/PreviewManager.cs]: `return m_NodesToUpdate.Count > 0;`

Minor tweak but more concise.

File: com.unity.shadergraph/Editor/Drawing/Views/GraphEditorView.cs [hunk 6]
```diff
@@ -459,7 +459,14 @@ public void UpdatePreviewShaders()
 
         public void HandleGraphChanges()
         {
-            previewManager.HandleGraphChanges();
+            if (previewManager.HandleGraphChanges() == true)
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Drawing/Views/GraphEditorView.cs]: Another tiny tweak but you don't need to compare to true inside the if.

**Pull request merged**